File: client/src/main/java/com/mechanitis/demo/client/RSocketClient.java
package com.mechanitis.demo.client;

import lombok.extern.log4j.Log4j2;
import org.springframework.messaging.rsocket.RSocketRequester;
import reactor.core.publisher.Flux;

@Log4j2
public class RSocketClient implements StockClient {
    private RSocketRequester rSocketRequester;

    RSocketClient(RSocketRequester rSocketRequester) {
        this.rSocketRequester = rSocketRequester;
    }

    @Override
    public Flux<StockPrice> pricesFor(String symbol) {
        log.info("RSocketClient");
        return rSocketRequester.route("stockPrices")
                               .data(symbol)
                               .retrieveFlux(StockPrice.class);
    }
}

File: client/src/main/kotlin/com/mechanitis/demo/client/kotlin/RSocketClient.kt
package com.mechanitis.demo.client.kotlin

import lombok.extern.log4j.Log4j2
import org.springframework.messaging.rsocket.RSocketRequester
import reactor.core.publisher.Flux

@Log4j2
class RSocketClient internal constructor(private val rSocketRequester: RSocketRequester) : StockClient {

    override fun pricesFor(symbol: String): Flux<StockPrice> {
        return rSocketRequester
            .route("stockPrices")
            .data(symbol)
            .retrieveFlux(StockPrice::class.java)
    }
}
